File: DataStructures/Stacks/atcoder/B51/B51-GPT.py
# 以下は、Python（CPython 3.11.4）での解答です。型アノテーションをすべて明示し、**Pylance** の型エラーが出ないように注意しています。

# また、処理時間・メモリの消費を抑えた構成で、**最大長20万の括弧列**にも耐えるようにしています。

# ---

# ## ✅ 概要

# * スタックを用いた **O(N)** の処理で対応括弧を収集。
# * 出力条件に合わせて **max(l, r)** 昇順にソート。
# * 処理はすべて関数化し、**型付きで記述**。

# ---

# ## ✅ 実装コード（Python 3.11.4）

# ```python
import sys
from typing import List, Tuple


def find_matching_parentheses(s: str) -> List[Tuple[int, int]]:
    """
    対応の取れた括弧列 s の各括弧の対応を求める。

    :param s: 対応の取れた括弧列（文字列）
    :return: 対応する括弧の位置ペア（1-indexed, 昇順ソート済み）
    """
    stack: List[int] = []
    pairs: List[Tuple[int, int]] = []

    for i, ch in enumerate(s):
        if ch == '(':
            stack.append(i + 1)  # 1-indexed
        else:
            if stack:
                left: int = stack.pop()
                pairs.append((left, i + 1))

    # 出力条件: max(l, r) 昇順
    pairs.sort(key=lambda x: max(x[0], x[1]))
    return pairs


def main() -> None:
    """
    標準入力を読み込み、対応括弧の位置を出力する。
    """
    s: str = sys.stdin.read().strip()
    result: List[Tuple[int, int]] = find_matching_parentheses(s)
    output: List[str] = [f"{l} {r}" for l, r in result]
    print("\n".join(output))


if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 処理時間・メモリ効率

# | 項目       | 計算量               | 詳細内容                         |
# | -------- | ----------------- | ---------------------------- |
# | 時間計算量    | `O(N log N)`      | N: 長さ。探索 O(N)、ソート O(N log N) |
# | 空間計算量    | `O(N)`            | スタック + 結果ペア（最大で N/2 個）       |
# | メモリ制限対応  | ✅ 最大 1024MiB に収まる |                              |
# | 実行時間制限対応 | ✅ 最大 2 秒以内に収まる    |                              |

# ---

# ## ✅ 型安全に関する注意点（Pylance 対応）

# * `List[int]` や `List[Tuple[int, int]]` を明示。
# * `.pop()` の戻り値は `int` として扱い、型推論が曖昧にならないようにしています。
# * `sys.stdin.read()` の戻り値を `str` と明記。

File: DataStructures/Stacks/atcoder/B51/B51-GPT.js
// 以下は、**対応の取れたカッコ列**に対して、全ての対応する括弧の位置を出力する **JavaScript（Node.js）** の実装です。

// * `fs` モジュールを用いて高速に標準入力を読み取り、
// * **スタックを使って対応関係を O(N)** で求め、
// * 出力条件（`max(li, ri)` 昇順）を満たすようにソートして出力します。

// ---

// ### ✅ 解法の概要（重要ポイント）:

// * `(` が来たら位置をスタックに積む。
// * `)` が来たらスタックから `(` の位置を取り出して対応関係として保存。
// * 最後に `max(l, r)` 昇順でソートして出力。

// ---

// ### ✅ 実装コード（Node.js 18.16.1, fs使用）

// ```javascript
// 標準入力の取得
const fs = require('fs');

/**
 * 入力文字列 S に対して、すべての括弧の対応を出力する
 * @param {string} S - 括弧列（対応の取れたカッコ列）
 * @returns {Array<[number, number]>} - 対応している括弧の位置ペア（1-indexed）
 */
function findMatchingBrackets(S) {
    const stack = [];
    const pairs = [];

    for (let i = 0; i < S.length; i++) {
        const ch = S[i];
        if (ch === '(') {
            stack.push(i + 1); // 1-indexed
        } else {
            const left = stack.pop();
            pairs.push([left, i + 1]);
        }
    }

    // 条件：max(l, r) 昇順でソート
    pairs.sort((a, b) => Math.max(a[0], a[1]) - Math.max(b[0], b[1]));

    return pairs;
}

// メイン処理
function main() {
    const S = fs.readFileSync('/dev/stdin', 'utf8').trim();
    const result = findMatchingBrackets(S);
    const output = result.map(pair => `${pair[0]} ${pair[1]}`).join('\n');
    console.log(output);
}

main();
// ```

// ---

// ### ✅ 処理時間・メモリ効率

// * **時間計算量:** O(N)（スタック操作 + ソート）
// * **空間計算量:** O(N)（スタック + 結果ペア）
// * **最大入力長 200,000 でも 2 秒 / 1024 MiB 制限内に収まります**

// ---

// 何か追加要件があれば教えてください（例えばペアを辞書順に並べたい、など）。
